File: py/sort_algorithm/Select_Sort.py
def SelectSort(a):
    # 先寻找最小的数, 然后交换最小值和首值, 以此类推
    n = len(a)
    for j in range(n - 1):
        # 遍历进行选择并交换
        k = j  # k是最小值元素的下标
        for i in range(j + 1, n):
            # 遍历寻找最小值对应的下标
            if a[k] > a[i]:
                k = i
        a[j], a[k] = a[k], a[j]
    return a


a = [54, 26, 93, 17, 77, 31, 44, 55, 20]

print(a)
print(SelectSort(a))

File: c_cpp/Sort/Select_Sort.cpp
#include <iostream>
#include <vector>

using namespace std;


void SelectSort1(vector<int> &arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; ++i) {
        int min1 = i;
        // unordered
        for (int j = i + 1; j < n; ++j)
            if (arr[j] < arr[min1]) min1 = j;
        // this location is not the min
        if (min1 != i) {
            /* swap(arr[min1], arr[i]); */
            int tmp = arr[min1];
            arr[min1] = arr[i];
            arr[i] = tmp;
        }
    }
}

void SelectSort2(vector<int> &arr) {
    int n = arr.size(), i{};
    while (i < n) {
        int min1 = i;
        for (int j = i + 1; j < n; ++j)
            if (arr[min1] > arr[j]) min1 = j;
        if (min1 != i) swap(arr[min1], arr[i]);
        i++;
    }
}

void SelectSort(vector<int> &arr) {
    int n = arr.size();
    for (int i{}; i < n - 1; ++i) {
        int min1 = i, j{i + 1};
        for (; j < n; ++j)
            if (arr[j] < arr[min1]) min1 = j;
        if (min1 != i) swap(arr[min1], arr[i]);
    }
}

// 此函数用于打印输出数组
void printArray(vector<int> arr) {
    for (size_t i = 0; i < arr.size(); ++i) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main(int argc, char const *argv[]) {
    // 定义数组
    vector<int> a = {19, 97, 9, 17, 1, 8};
    // 输出原始数组
    printArray(a);
    // 排序
    SelectSort(a);
    // 输出排序后的数组
    printArray(a);
    return 0;
}

